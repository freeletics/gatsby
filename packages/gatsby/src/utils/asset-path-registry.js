const path = require(`path`)
const fs = require(`fs-extra`)

/*
 * TODO: persist this between runs?
 */
let assets = new Set()

exports.register = function register(file) {
  if (file) {
    assets.add(file)
  }
  return assets
}

exports.remove = function remove(file) {
  assets.delete(file)
  return assets
}

exports.clear = function clear() {
  assets.clear()

  return assets
}

exports.getAssets = async function getAssets(directory) {
  const webpackStats = path.join(directory, `public`, `webpack.stats.json`)
  let webpackAssets
  try {
    const { namedChunkGroups = {} } = await fs
      .readFile(webpackStats, `utf8`)
      .then(data => JSON.parse(data))

    webpackAssets = Object.keys(namedChunkGroups).reduce(
      (merged, chunkName) => {
        const group = namedChunkGroups[chunkName]
        group.assets.forEach(asset => {
          merged.add(asset)
        })
        return merged
      },
      new Set()
    )
  } catch (e) {
    webpackAssets = new Set()
  }

  return new Set([...assets, ...webpackAssets])
}
