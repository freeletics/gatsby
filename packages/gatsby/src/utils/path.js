const path = require(`path`)
const os = require(`os`)

const { register } = require(`./asset-path-registry`)

function joinPath(...paths) {
  const joinedPath = path.join(...paths)
  if (os.platform() === `win32`) {
    return joinedPath.replace(/\\/g, `\\\\`)
  } else {
    return joinedPath
  }
}

function withBasePath(basePath) {
  return (...paths) => joinPath(basePath, ...paths)
}

function withPrefix(prefix, callback) {
  return function(...parts) {
    const filePath = path.join(prefix, ...parts)
    if (typeof callback === `function`) {
      callback(filePath)
    }
    return filePath
  }
}

exports.joinPath = joinPath
exports.withBasePath = withBasePath
exports.withPathPrefix = withPrefix

exports.withAssetPrefix = prefix => withPrefix(prefix, register)
