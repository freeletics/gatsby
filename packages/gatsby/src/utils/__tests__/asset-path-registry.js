import path from "path"
import { register, clear, remove, getAssets } from "../asset-path-registry"

beforeEach(clear)

describe(`general registry behavior`, () => {
  it(`starts with empty registry`, () => {
    const registry = register()
    expect(registry.size).toBe(0)
  })

  it(`registers a file path to registry`, () => {
    const registry = register(`a.js`)
    expect(registry.size).toBe(1)
  })

  it(`can remove a file path`, () => {
    const file = `a.js`
    register(file)
    const registry = remove(file)

    expect(registry.size).toBe(0)
  })

  it(`does not add same file path twice`, () => {
    const file = `a.js`
    new Array(10).fill(undefined).forEach(() => register(file))

    expect(register().size).toBe(1)
  })
})

describe(`getAssets`, () => {
  it(`returns in-memory assets`, async () => {
    const files = [`a.js`, `b.js`]
    files.forEach(file => register(file))
    const assets = await getAssets(path.join(__dirname, `__not_a_real_path__`))

    expect(Array.from(assets)).toEqual(files)
  })

  it(`returns webpack assets`, async () => {
    const assets = await getAssets(path.join(__dirname, `fixtures`))

    expect(assets).toMatchSnapshot()
  })

  it(`returns webpack and in-memory assets`, async () => {
    const files = [`a.js`, `b.js`]
    files.forEach(file => register(file))

    const assets = await getAssets(path.join(__dirname, `fixtures`))
    files.forEach(file => {
      expect(assets.has(file)).toBe(true)
    })

    expect(assets.size - files.length).not.toBe(0)
  })
})
