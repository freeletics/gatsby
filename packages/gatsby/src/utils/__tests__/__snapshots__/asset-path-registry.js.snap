// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAssets returns webpack assets 1`] = `
Set {
  "webpack-runtime-f08922a061a7e09baa26.js",
  "webpack-runtime-f08922a061a7e09baa26.js.map",
  "app-4b433d29b46598f0ecd3.js",
  "app-4b433d29b46598f0ecd3.js.map",
  "component---node-modules-gatsby-plugin-offline-app-shell-js-b1a75793f38b5d3df59f.js",
  "component---node-modules-gatsby-plugin-offline-app-shell-js-b1a75793f38b5d3df59f.js.map",
  "0.2b5955a752ed0afe77ce.css",
  "0-fecec88f7d9d61b9013a.js",
  "0-fecec88f7d9d61b9013a.js.map",
  "1-b047c021df576f0acceb.js",
  "1-b047c021df576f0acceb.js.map",
  "2.fd4497ec960200ce8103.css",
  "2-5ce6e769c43c7185e581.js",
  "2-5ce6e769c43c7185e581.js.map",
  "3-acdb3c000e2ba39d41ef.js",
  "3-acdb3c000e2ba39d41ef.js.map",
  "component---src-templates-template-blog-list-js-e53084eb5bee3f6f7f3e.js",
  "component---src-templates-template-blog-list-js-e53084eb5bee3f6f7f3e.js.map",
  "40-cfb9271892b3c6f32cb7.js",
  "40-cfb9271892b3c6f32cb7.js.map",
  "component---src-templates-template-blog-post-js-7b76cbf47a3d2298e372.js",
  "component---src-templates-template-blog-post-js-7b76cbf47a3d2298e372.js.map",
  "component---src-templates-tags-js-596abb26f29b385da04f.js",
  "component---src-templates-tags-js-596abb26f29b385da04f.js.map",
  "component---src-templates-template-starter-page-js-3ffff961577834cf4023.js",
  "component---src-templates-template-starter-page-js-3ffff961577834cf4023.js.map",
  "component---src-templates-template-contributor-page-js-f9404cf60eca8d73fa02.js",
  "component---src-templates-template-contributor-page-js-f9404cf60eca8d73fa02.js.map",
  "component---src-templates-template-creator-details-js-88ef970e91e0b78c2f36.js",
  "component---src-templates-template-creator-details-js-88ef970e91e0b78c2f36.js.map",
  "43-53d7dbdaf8c4c4002036.js",
  "43-53d7dbdaf8c4c4002036.js.map",
  "component---src-templates-template-showcase-details-js-f5e8e7b262e87f30a837.js",
  "component---src-templates-template-showcase-details-js-f5e8e7b262e87f30a837.js.map",
  "component---src-templates-template-docs-markdown-js-beed50a1f7dd4d27b797.js",
  "component---src-templates-template-docs-markdown-js-beed50a1f7dd4d27b797.js.map",
  "4-a765cc45c3be4b896d6f.js",
  "4-a765cc45c3be4b896d6f.js.map",
  "6-877482b381b70a3e1930.js",
  "6-877482b381b70a3e1930.js.map",
  "component---src-templates-template-docs-local-packages-js.fd4497ec960200ce8103.css",
  "component---src-templates-template-docs-local-packages-js-c53e4baf78ad21580235.js",
  "component---src-templates-template-docs-local-packages-js-c53e4baf78ad21580235.js.map",
  "component---src-templates-template-docs-remote-packages-js.fd4497ec960200ce8103.css",
  "component---src-templates-template-docs-remote-packages-js-68b7685d85bd50607d49.js",
  "component---src-templates-template-docs-remote-packages-js-68b7685d85bd50607d49.js.map",
  "component---src-pages-404-js-c31dd87a2028cf26b810.js",
  "component---src-pages-404-js-c31dd87a2028cf26b810.js.map",
  "component---src-pages-blog-tags-js-46843cd191f6d9a81483.js",
  "component---src-pages-blog-tags-js-46843cd191f6d9a81483.js.map",
  "5-5f0403c3cf66cc31c4fb.js",
  "5-5f0403c3cf66cc31c4fb.js.map",
  "component---src-pages-community-agencies-js-442a248e84f48ec18e11.js",
  "component---src-pages-community-agencies-js-442a248e84f48ec18e11.js.map",
  "component---src-pages-community-companies-js-4a9bb59d522d16bc0cd5.js",
  "component---src-pages-community-companies-js-4a9bb59d522d16bc0cd5.js.map",
  "component---src-pages-community-index-js-3c9cc70d04f424db3ffa.js",
  "component---src-pages-community-index-js-3c9cc70d04f424db3ffa.js.map",
  "component---src-pages-community-people-js-26220872f1564524ef20.js",
  "component---src-pages-community-people-js-26220872f1564524ef20.js.map",
  "component---src-pages-docs-actions-js-af5b12753a00718765fd.js",
  "component---src-pages-docs-actions-js-af5b12753a00718765fd.js.map",
  "component---src-pages-docs-browser-apis-js-f9529c1875808d214471.js",
  "component---src-pages-docs-browser-apis-js-f9529c1875808d214471.js.map",
  "component---src-pages-docs-index-js-c7323417d5798f1bcd94.js",
  "component---src-pages-docs-index-js-c7323417d5798f1bcd94.js.map",
  "component---src-pages-docs-node-apis-js-18a7ea5c808fce0d9402.js",
  "component---src-pages-docs-node-apis-js-18a7ea5c808fce0d9402.js.map",
  "component---src-pages-docs-ssr-apis-js-c4b332ac75f8093bdf42.js",
  "component---src-pages-docs-ssr-apis-js-c4b332ac75f8093bdf42.js.map",
  "component---src-pages-features-js-4a4a3f05f2ed8a3afa03.js",
  "component---src-pages-features-js-4a4a3f05f2ed8a3afa03.js.map",
  "component---src-pages-newsletter-js-b823f6af0f63d49f5bef.js",
  "component---src-pages-newsletter-js-b823f6af0f63d49f5bef.js.map",
  "component---src-pages-plugins-js.fd4497ec960200ce8103.css",
  "component---src-pages-plugins-js-259ea07a4bf263d315c7.js",
  "component---src-pages-plugins-js-259ea07a4bf263d315c7.js.map",
  "42-d5c0588bf92953044654.js",
  "42-d5c0588bf92953044654.js.map",
  "component---src-pages-showcase-js-710ac9cbe8f04ccc5629.js",
  "component---src-pages-showcase-js-710ac9cbe8f04ccc5629.js.map",
  "39-72be356c49b30af6ec24.js",
  "39-72be356c49b30af6ec24.js.map",
  "component---src-pages-starters-js-5cea20767c8766bdcda4.js",
  "component---src-pages-starters-js-5cea20767c8766bdcda4.js.map",
  "component---src-pages-ecosystem-js-633f4f4caa8ef50afcbe.js",
  "component---src-pages-ecosystem-js-633f4f4caa8ef50afcbe.js.map",
  "component---src-pages-index-js-0a2a452fc816778a6329.js",
  "component---src-pages-index-js-0a2a452fc816778a6329.js.map",
}
`;
